velox/dwio/parquet/reader/ParquetReader.cpp

```diff
@@ -279,24 +279,42 @@ std::shared_ptr<const ParquetTypeWithId> ReaderBase::getParquetColumnInfo(
               maxDefine);
         }
         case thrift::ConvertedType::MAP_KEY_VALUE: {
-          // child of MAP
-          VELOX_CHECK_EQ(
-              schemaElement.repetition_type,
-              thrift::FieldRepetitionType::REPEATED);
-          assert(children.size() == 2);
-          auto childrenCopy = children;
-          return std::make_shared<const ParquetTypeWithId>(
-              TypeFactory<TypeKind::MAP>::create(
-                  children[0]->type(), children[1]->type()),
-              std::move(childrenCopy),
-              curSchemaIdx, // TODO: there are holes in the ids
-              maxSchemaElementIdx,
-              ParquetTypeWithId::kNonLeaf, // columnIdx,
-              std::move(name),
-              std::nullopt,
-              std::nullopt,
-              maxRepeat,
-              maxDefine);
+          if (children.size() == 1) {
```

Comment:
@hitarth I don't think this fix is very accurate. According to the spec, the rule to test the backward compatibility is "a group annotated with MAP_KEY_VALUE that is not contained by a MAP-annotated group should be handled as a MAP-annotated group.". You're checking its children count, not if its parent is a MAP node. This may somewhat work for the map type, but won't work for list. 

I think the correct fix would be following exactly the rules defined in the spec, i.e. check its parent node's type. If it's parent is not MAP, then it would be treated as MAP instead of MAP_KEY_VALUE. For that, we will need to pass in the parent schemaIdx. We can add a new parameter "uint32_t parentSchemaIdx" so that when a node is MAP_KEY_VALUE, we can check its parent type by schema[parentSchemaIdx].converted_type. Similarly we can do the same for list. Will you be able to try this approach to see if it works?
Please also try to reduce the amount of repeated code in your new rewrite.

Reply:
Thanks for the suggestion, let me try adding parentSchemaIdx and rewrite this.